oak-run/src/main/java/org/apache/jackrabbit/oak/indexversion/PurgeOldIndexVersion.java

```diff
@@ -0,0 +1,226 @@
+                if (isReadWriteRepository) {
+                    purgeOldIndexVersion(nodeStore, toDeleteIndexNameObjectList, trimmedBaseIndexPath);
+                } else {
+                    log.info("Repository is opened in read-write mode: IndexOperations" +
+                            " for index at path {} are : {}", trimmedBaseIndexPath, toDeleteIndexNameObjectList);
+                }
+            }
+        }
+    }
+
+    private Map<String, Set<String>> segregateIndexes(Set<String> indexPathSet) {
+        Map<String, Set<String>> segregatedIndexes = new HashMap<>();
+        for (String path : indexPathSet) {
+            String baseIndexPath = getBaseIndexPath(path);
+            Set<String> indexPaths = segregatedIndexes.get(baseIndexPath);
+            if (indexPaths == null) {
+                indexPaths = new HashSet<>();
+            }
+            indexPaths.add(PurgeOldVersionUtils.trimSlash(path));
+            segregatedIndexes.put(baseIndexPath, indexPaths);
+        }
+        return segregatedIndexes;
+    }
+
+    private Iterable<String> getIndexPathList(NodeStore store) throws CommitFailedException, IOException {
+        IndexPathService indexPathService = new IndexPathServiceImpl(store);
+        Iterable<String> indexPaths = indexPathService.getIndexPaths();
+        return indexPaths;
+    }
+
+    private boolean isIndexPathOptionValid(String indexPathOption, String indexPath) {
+        String slashTrimmedIndexPathOption = PurgeOldVersionUtils.trimSlash(indexPathOption);
+        String slashTrimmedIndexPath = PurgeOldVersionUtils.trimSlash(indexPath);
+        if (slashTrimmedIndexPath.startsWith(slashTrimmedIndexPathOption)) {
+            if (slashTrimmedIndexPathOption.length() == slashTrimmedIndexPath.length()
+                    || (slashTrimmedIndexPathOption.length() < slashTrimmedIndexPath.length()
+                    && (slashTrimmedIndexPath.charAt(slashTrimmedIndexPathOption.length()) == '/'
+                    || slashTrimmedIndexPath.charAt(slashTrimmedIndexPathOption.length()) == '-'))) {
+                return true;
+            } else {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @param indexPathList: list of indexpaths retrieved from  index service
+     * @return returns set of indexpaths which are to be considered for purging
+     */
+    private Set<String> filterIndexPaths(Iterable<String> indexPathList) {
+        Set<String> filteredIndexPaths = new HashSet<>();
+        for (String indexPathOption : this.indexPaths) {
+            for (String indexPath : indexPathList) {
+                if (isIndexPathOptionValid(indexPathOption, indexPath)) {
+                    filteredIndexPaths.add(PurgeOldVersionUtils.trimSlash(indexPath));
+                }
+            }
+        }
+        return filteredIndexPaths;
+    }
+
+    StringBuilder stringJoiner(String[] stringArray, int arrayIndexToJoin, String delimiter) {
```

Comment:
I would prefer not having this method and use PathUtils instead... if we need it, then maybe move to a utility method, add javadocs, make it static, and add unit tests.

Reply:
Changed to PathUtils.